frontend/src/modules/dataset/dataset-browser.tsx

```diff
@@ -0,0 +1,389 @@
+            filteringTextDisplay: '',
+            filteredItems: []
+        },
+        pagination: {
+            currentPageIndex: 1,
+            pagesCount: 1,
+            openEnd: false,
+            disabled: false,
+        },
+        refresh: false
+    });
+
+    const displayMode = displayModeForDatasetContext(state.datasetContext);
+    const isCreating = state.manageMode === DatasetBrowserManageMode.Create;
+
+    // memoize notification service so it doesn't cause infinite loops for fetchDatasetResources & fetchDatasets
+    const notificationService = useMemo(() => {
+        return NotificationService(dispatch);
+    }, [dispatch]);
+
+    const fetchDatasetResources = useMemo(() => (datasetContext: Partial<DatasetContext>, nextToken?: string, existingItems?: (DatasetResource | IDataset)[]) => {
+        if (state.manageMode !== DatasetBrowserManageMode.Create) {
+            return dispatch(getDatasetContents({datasetContext, projectName, username, nextToken})).then((response) => {
+                if (isFulfilled(response)) {
+                    // if nextToken is provided we're appending to existing items
+                    const items = nextToken ? [...existingItems || []] : [];
+        
+                    const payload: Partial<Pick<DatasetBrowserState, 'items' | 'nextToken' | 'datasetContext' | 'filter' | 'isLoading' | 'refresh'>> = {
+                        items: items.concat(response.payload.data.contents),
+                        nextToken: response.payload.data.nextToken,
+                        isLoading: false
+                    };
+        
+                    // if the dataset context location is for a resource update the filter too
+                    const resourceStringForPath = resourceForPath(datasetContext.location);
+                    if (resourceStringForPath) {
+                        payload.filter = {
+                            filteredItems: [],
+                            filteringTextDisplay: resourceStringForPath,
+                            filteringText: resourceStringForPath
+                        };
+                    }
+        
+                    setState({
+                        type: DatasetActionType.State,
+                        payload
+                    });
+                } else {
+                    // if an error occurred reset the component to a safe state
+                    notificationService.generateNotification(`Failed to fetch resources for ${datasetContext.name}`, 'error');
+                    setState({
+                        type: DatasetActionType.DatasetContext,
+                        payload: undefined
+                    });
+                }
+            });
+        }
+
+        setState({
+            type: DatasetActionType.State,
+            payload: {isLoading: false}
+        });
+    }, [dispatch, projectName, username, notificationService, state.manageMode]);
+
+    const fetchDatasets = useMemo(() => () => {
```

Comment:
You're seeing it more here because this component stores a lot more state stored internally instead of in the application state. This is a more full-fleged component that has a good amount of temporary state that doesn't make sense to store in the application store.